### PR TITLE
minion_id_remove_domain could contain string and bool values

### PR DESCRIPTION
**Pull request by marbx**
This PR replaces PR https://github.com/saltstack/salt/pull/49213

## What does this PR do?

When a minion lacks its minion_id at startup, salt.config.\_\_init\_\_.py requests one from utils.network.generate_minion_id(), which returns the fully qualified domain name (FQDN) of the host.

When all minions are within one domain and named after their hostnames, generated a "FQDN minion id" would be an inconsistent naming.

salt.config.\_\_init\_\_.py has the option to lowercase the minion_id.
salt.config.\_\_init\_\_.py also has the option to append a domain to a generated "hostname minion id".

I hereby propose a new option, to generate a "hostname minion id", by removing a single domain from a FQDN.

## What issues does this PR fix or reference?

Issue #49212

## Previous Behavior (example)

The generated minion_id is king_bob.foo.org

## New Behavior (example)

With `minion_id_remove_domain: False`, as previously.
With `minion_id_remove_domain: bar.org`, as previously.
With `minion_id_remove_domain: foo.org`, the generated minion_id becomes king_bob
With `minion_id_remove_domain: True`, the generated minion_id becomes king_bob

## Tests written?

Yes

### CODE REVIEW

File: salt/config/__init__.py [hunk 1]
```diff
@@ -944,6 +944,9 @@ def _gather_buffer_space():
     # Always generate minion id in lowercase.
     'minion_id_lowercase': bool,
 
+    # Remove this single domain (e.g. 'foo.org') in a generated minion id.
+    'minion_id_remove_domain': six.string_types,
```
Comment (DmitryKuzmenko) [inline salt/config/__init__.py]:
```
 'minion_id_remove_domain': (six.string_types, bool)
```

File: salt/config/__init__.py [hunk 2]
```diff
@@ -1453,6 +1456,7 @@ def _gather_buffer_space():
     'grains_refresh_every': 0,
     'minion_id_caching': True,
     'minion_id_lowercase': False,
+    'minion_id_remove_domain': '',
```
Comment (DmitryKuzmenko) [inline salt/config/__init__.py]:
```
'minion_id_remove_domain': False
```

File: salt/config/__init__.py [hunk 3]
```diff
@@ -3649,6 +3653,21 @@ def get_id(opts, cache_minion_id=False):
     if opts.get('minion_id_lowercase'):
         newid = newid.lower()
         log.debug('Changed minion id %s to lowercase.', newid)
+
+    # Optionally remove a domain or many in a generated minion id
+    if opts.get('minion_id_remove_domain'):
+        rem_domain = opts.get('minion_id_remove_domain')
```
Comment (DmitryKuzmenko) [inline salt/config/__init__.py]: Better is either:
```
￼    remove_domain = opts.get('minion_id_remove_domain')
    if remove_domain:
```
or
```
    if opts.get('minion_id_remove_domain'):
￼         remove_domain = opts.get['minion_id_remove_domain']
```

File: salt/config/__init__.py [hunk 4]
```diff
@@ -3649,6 +3653,21 @@ def get_id(opts, cache_minion_id=False):
...
+        if isinstance(rem_domain, bool):
```
Comment (DmitryKuzmenko) [inline salt/config/__init__.py]:
```
if rem_domain is True:
```

**Review by DmitryKuzmenko — CHANGES_REQUESTED** [review]
LGTM, I'd just like to see some minor changes in there.

File: salt/config/__init__.py (hunk 1 above)
Comment (markuskramerIgitt) [reply salt/config/__init__.py]: ok

File: salt/config/__init__.py (hunk 2 above)
Comment (markuskramerIgitt) [reply salt/config/__init__.py]: ok

File: salt/config/__init__.py (hunk 4 above)
Comment (markuskramerIgitt) [reply salt/config/__init__.py]: ok

File: salt/config/__init__.py (hunk 1 above)
Comment (isbm) [reply salt/config/__init__.py]: @markuskramerIgitt it is just a list of _possible_ types. So just like @DmitryKuzmenko suggested.

File: salt/config/__init__.py (hunk 4 above)
Comment (isbm) [reply salt/config/__init__.py]: @markuskramerIgitt This is actually important for you to know in the future: you're only looking if `rem_domain` is a type of Boolean, but you do not know if it is True or False. Also, if you would just ask `if rem_domain:` then you get back a Boolean, and if it is True, still don't know what to do: remove specific domain, or any. Hence in these kind of situation we always explicitly looking for both type and value at the same time: `is None`, `is not None`, `is True` etc.

Hope it helps. :wink:

File: doc/ref/configuration/minion.rst [hunk 5]
```diff
@@ -635,6 +635,38 @@ FQDN (for instance, Solaris).
...
+when the minions shall be named like hostnames. Can be a single domain (to
+prevent name clashes), or True, to remove all domains.
+
+Examples:
+ - minion_id_remove_domain = foo.org
+   - FQDN = king_bob.foo.org --> minion_id = king_bob
+   - FQDN = king_bob.bar.org --> minion_id = king_bob.bar.org
+ - minion_id_remove_domain = True
+   - FQDN = king_bob.foo.org --> minion_id = king_bob
+   - FQDN = king_bob.bar.org --> minion_id = king_bob
```
Comment (isbm) [inline doc/ref/configuration/minion.rst]: Maybe rephrase this by just plain English, something like:

When `minion_id_remove_domain` is set to `foo.org`, then only this domain will be removed from the minion ID. For example, `king_bob.foo.org` will be `king_bob` and `king_bob.bar.org` will remain unchanged. In case `minion_id_remove_domain` is set to `True`, then all domains will be removed. For example, both `king_bob.foo.org` and `king_bob.bar.org` will be `king_bob`.

How's that sounds like? Feel free to adapt it and change it.

**Review by isbm — APPROVED** [review]
I am approving this, thanks @markuskramerIgitt ! However, I have just one suggestion to rephrase your table-list schematic example into something generic, typically readable. I include my suggestion.

File: doc/ref/configuration/minion.rst (hunk 5 above)
Comment (markuskramerIgitt) [reply doc/ref/configuration/minion.rst]: I prefer table-list like schematic examples, also here.

**Comment by markuskramerIgitt** [comment]
Hello @isbm, the page states you approved my changes. Why does the page still shows "Changes requested"?

File: doc/ref/configuration/minion.rst [hunk 6]
```diff
@@ -635,6 +635,38 @@ FQDN (for instance, Solaris).
 
     append_domain: foo.org
 
+.. conf_minion:: minion_id_remove_domain
+
+``minion_id_remove_domain``
+---------------------
+
+.. versionadded:: 18.3.?
```
Comment (terminalmage) [inline doc/ref/configuration/minion.rst]: The version here should be ``Neon``. We will replace with the correct version once we prepare that release.

File: salt/config/__init__.py [hunk 7]
```diff
@@ -3649,6 +3653,21 @@ def get_id(opts, cache_minion_id=False):
...
         log.debug('Changed minion id %s to lowercase.', newid)
+
+    # Optionally remove a domain or many in a generated minion id
+    rem_domain = opts.get('minion_id_remove_domain')
+    if rem_domain:
+        if rem_domain is True:
+            if '.' in newid:
+                # Clean away any domain
+                newid, xdomain = newid.split('.', 1)
+                log.debug('Cleaned away any domain (%s) from minion id.', xdomain)
```
Comment (terminalmage) [inline salt/config/__init__.py]: "Cleaned away" doesn't make a lot of sense here, you can use "Removed" as in the log message you added below.

**Review by terminalmage — CHANGES_REQUESTED** [review]
We need test cases for the `True` and `False` values for this config option. Also, is this intended to replace https://github.com/saltstack/salt/pull/49213? If so, we should probably close https://github.com/saltstack/salt/pull/49213.

**Comment by isbm** [comment]
@markuskramerIgitt because @terminalmage needs unit tests (+1, BTW)  and some minor fixes. :wink:

**Comment by terminalmage** [comment]
Actually, I think it is because @DmitryKuzmenko had requested changes and has not re-reviewed with an `Approved` outcome.

**Review by DmitryKuzmenko — APPROVED** [review]

**Comment by DmitryKuzmenko** [comment]
who? me? sorry, done. =-/

**Comment by markuskramerIgitt** [comment]
Of the two PR...
  - #49213 is string type and  complete
 - #49378 is (string,bool) type, needs refactoring due to cyclomatic complexity 22 in get_id(), boolean test cases and a more serene description for boolean True.

… only one is necessary (they are mutually exclusive).
Either one is fine for me

 @terminalmage, which do you like? Can you decide alone?

@isbm, if #49378 is favored, could you please change/add the Boolean part? I see that you have more experience in Python.

**Comment by isbm** [comment]
@markuskramerIgitt what exactly should I change?

**Comment by terminalmage** [comment]
I like this approach better than the other PR, but the changes I requested still need to be made before this PR is ready for merge.

File: doc/ref/configuration/minion.rst (hunk 6 above)
Comment (markuskramerIgitt) [reply doc/ref/configuration/minion.rst]: Aah, I first read `None`... 
Set to `Neon`.

I will backport this feature into the msi packaged 18.3.2 because we need it soon.

**Comment by markuskramerIgitt** [comment]
@terminalmage, thank you for your vote, I see there are more cases for "mixed type" in the same module. Have I now  completed this PR?

I think both PR #49213 and #49378 are ready for merge.
Please accept one and close the other.

File: salt/config/__init__.py (hunk 2 above)
Comment (markuskramerIgitt) [reply salt/config/__init__.py]: Hi @DmitryKuzmenko, thank you for your suggestions. 
I never used (string, bool) before. 
Already checked in.

I will "resolve" this conversation because GitHub still sees open "requested changes".

File: salt/config/__init__.py (hunk 3 above)
Comment (markuskramerIgitt) [reply salt/config/__init__.py]: I use your second proposal. 

I also had to move the code into a separate function because it increased `get_id()` to a  cyclomatic complexity of 22.

File: salt/config/__init__.py (hunk 4 above)
Comment (markuskramerIgitt) [reply salt/config/__init__.py]: It is the first time that I make use of `is True`, thank you for your explanation, @isbm.

I will "resolve" this conversation because GitHub sees open "requested changes".

**Comment by markuskramerIgitt** [comment]
Hi @terminalmage, above it says that you "request changes", but I don't find which ones.
Which changes do you request?

I also don't understand which checks exactly are "not successful".

File: salt/config/__init__.py (hunk 7 above)
Comment (markuskramerIgitt) [reply salt/config/__init__.py]: Hi @terminalmage, I implemented the changes you requested (test cases for True and False, closed PR 49213 and changed the wording). 

Could you please comment or approve?

**Review by terminalmage — CHANGES_REQUESTED** [review]
@markuskramerIgitt The changes I requested were from my previous review, including adding tests. Those have been done, but there are some changes that need to be made to the docs. Once they are made this will be ready to merge.

File: doc/ref/configuration/minion.rst [hunk 8]
```diff
@@ -635,6 +635,38 @@ FQDN (for instance, Solaris).
 
     append_domain: foo.org
 
+.. conf_minion:: minion_id_remove_domain
+
+``minion_id_remove_domain``
+---------------------
```
[the remark this answers is not in the document]
Comment (markuskramerIgitt) [reply doc/ref/configuration/minion.rst]: Hi @terminalmage, thank you for the explanation of your change requests. 
I was not aware of Sphinx and  I  will gladfully adopt these rules.

Fixed in  https://github.com/saltstack/salt/pull/49378/commits/f5d6157257ccb51934d3496a5284ef403e55e614

File: doc/ref/configuration/minion.rst [hunk 9]
```diff
@@ -635,6 +635,38 @@ FQDN (for instance, Solaris).
...
+Examples:
+ - minion_id_remove_domain = foo.org
+   - FQDN = king_bob.foo.org --> minion_id = king_bob
+   - FQDN = king_bob.bar.org --> minion_id = king_bob.bar.org
+ - minion_id_remove_domain = True
+   - FQDN = king_bob.foo.org --> minion_id = king_bob
+   - FQDN = king_bob.bar.org --> minion_id = king_bob
+
+
+For more information, please see `Issue #49212`_ and `Pull Request #49378`_.
```
[the remark this answers is not in the document]
Comment (markuskramerIgitt) [reply doc/ref/configuration/minion.rst]: As above,  fixed in  https://github.com/saltstack/salt/pull/49378/commits/f5d6157257ccb51934d3496a5284ef403e55e614

File: doc/ref/configuration/minion.rst [hunk 10]
```diff
@@ -635,6 +635,38 @@ FQDN (for instance, Solaris).
...
+
+.. code-block:: yaml
+
+    minion_id_remove_domain: foo.org
+
+.. _Issue #49212: https://github.com/saltstack/salt/issues/49212
+.. _Pull Request #49378: https://github.com/saltstack/salt/pull/49378
```
[the remark this answers is not in the document]
Comment (markuskramerIgitt) [reply doc/ref/configuration/minion.rst]: As above,  fixed in  https://github.com/saltstack/salt/pull/49378/commits/f5d6157257ccb51934d3496a5284ef403e55e614

File: doc/ref/configuration/minion.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (markuskramerIgitt) [reply doc/ref/configuration/minion.rst]: the pull request mentions the issue, which only is a stub anyhow.

**Comment by markuskramerIgitt** [comment]
Hi @terminalmage , is there something I can do here?

File: doc/ref/configuration/minion.rst [hunk 11]
```diff
@@ -658,15 +658,12 @@ Examples:
    - FQDN = king_bob.bar.org --> minion_id = king_bob
 
 
-For more information, please see `Issue #49212`_ and `Pull Request #49378`_.
+For more information, please see :pull:`49378`.
```
Comment (terminalmage) [inline doc/ref/configuration/minion.rst]: Why was the issue number removed here, instead of using the `:issue:` RST role?

Comment (markuskramerIgitt) [reply doc/ref/configuration/minion.rst]: For no good reason. I just commited the `:issue:` RST role

**Comment by markuskramerIgitt** [comment]
Codeclimate fails because  

Function `remove_domain_from_fqdn` has a Cognitive Complexity of 6 (exceeds 5 allowed). Consider refactoring.

Do you really want to limit the Cognitive Complexity of a function at 5?

**Comment by rallytime** [comment]
@markuskramerIgitt This is fine. We need to do some tuning there.

**Comment by markuskramerIgitt** [comment]
Thank you, @rallytime.

**Comment by isbm** [comment]
@markuskramerIgitt Code Climate is to suggest only. And it is not yet there to be tuned. Lint is to block though. :wink:

**Comment by markuskramerIgitt** [comment]
Hi @terminalmage, do you have more changes for me?

**Review by terminalmage — APPROVED** [review]

**Comment by markuskramerIgitt** [comment]
Hi @terminalmage, thank you for approval.

**Comment by rallytime** [comment]
@markuskramerIgitt There are some test failures here: https://jenkinsci.saltstack.com/job/pr-kitchen-ubuntu1604-py2/job/PR-49378/4/

Can you take another look?

**Comment by markuskramerIgitt** [comment]
Hi @rallytime, 
The offending lines 719 and 753  in  `tests/unit/test_config.py` are both:

    config = sconfig.minion_config(minion_config)

and there are 5 more occasions in the same file.

The line reads a minion-configuration from file and is part of the test framework, not a part of testing something in specific .

I try to understand why this raises `TypeError: unicode argument without an encoding` so far no idea.

Could someone with knowledge of the testing framework have a look too?

**Comment by markuskramerIgitt** [comment]
Hi @isbm,
the above exception is thrown by [salt/_compat.py#L194](https://github.com/saltstack/salt/blob/a5b95787de9007947bad0a098e7ce63d50d22dac/salt/_compat.py#L194)

            try:
                packed = bool(int(str(bytearray(data)).encode('hex'), 16))
            except ValueError:

This is intended as a test if data is a packed binay.
I would  add TypeError to the known exception list.

10 days ago you altered this line - what do you think?

Thank you,
Markus

**Comment by isbm** [comment]
@markuskramerIgitt hmm, why do we have there TypeError? What you're getting into `data` variable then?

**Comment by isbm** [comment]
@markuskramerIgitt oh OK, I know why. You found a bug, congrats. PR to fix this: https://github.com/saltstack/salt/pull/49908

**Comment by rallytime** [comment]
Nice - thanks for helping out here @isbm!

**Comment by markuskramerIgitt** [comment]
Hi @isbm, thank you for picking it up.

**Comment by markuskramerIgitt** [comment]
Jenkins  still fails, but for no (apparent) reason?

Is there something I can do here?

**Comment by rallytime** [comment]
@markuskramerIgitt We're working through an issue with our pip installer right now, so no need to worry here right now. We'll get these going again when the issue is resolved. Thanks!

**Comment by markuskramerIgitt** [comment]
@rallytime, thank you for merging.
How will this pull request make its way from 'develop' into a release?

**Comment by gtmanfred** [comment]
This will be in the Neon feature release.

**Comment by markuskramerIgitt** [comment]
Thank you, now I unterstand `versionadded:: Neon`.
[Neon comes after Fluorine](https://github.com/saltstack/salt/blob/develop/doc/topics/releases/version_numbers.rst#code-names)

**Comment by markuskramerIgitt** [comment]
@gtmanfred, could you please consider to merge this change into a Fluorine release?
Otherwise I would need to create a local modification.

**Comment by markuskramerIgitt** [comment]
@gtmanfred, please forget to consider merging at an earlier time.
I just found that I can help myself with `id_function`, no local modification needed.

**Comment by markuskramerIgitt** [comment]
As of today, Neon still has no release version.

**Comment by markuskramerIgitt** [comment]
> @gtmanfred, please forget to consider merging at an earlier time.
> I just found that I can help myself with `id_function`, no local modification needed.

Unfortunately: `id_function`, requries local modification of the salt master or modification of the salt-minion installer.

The installers would have to  configure `id_function` before the minion service starts.

**Comment by markuskramerIgitt** [comment]
`minion_id_remove_domain`  is not in git checkout v2019.2.1  and git checkout v2019.2.3 but in 

git checkout v2019.8

git checkout v3000
